transport/server.cc

```diff
@@ -976,12 +976,12 @@ template <typename Process>
                                    cql3::computed_function_values,
                                    cql3::dialect>
 future<cql_server::result_with_foreign_response_ptr>
-cql_server::connection::process_on_shard(::shared_ptr<messages::result_message::bounce_to_shard> bounce_msg, uint16_t stream, fragmented_temporary_buffer::istream is,
-        service::client_state& cs, service_permit permit, tracing::trace_state_ptr trace_state, cql3::dialect dialect, Process process_fn) {
-    return _server.container().invoke_on(*bounce_msg->move_to_shard(), _server._config.bounce_request_smp_service_group,
-            [this, is = std::move(is), cs = cs.move_to_other_shard(), stream, permit = std::move(permit), process_fn,
+cql_server::connection::process_on_shard(shard_id shard, uint16_t stream, fragmented_temporary_buffer::istream is,
+        service::client_state& cs, service_permit permit, tracing::trace_state_ptr trace_state, cql3::dialect dialect, cql3::computed_function_values&& cached_vals, Process process_fn) {
+    return _server.container().invoke_on(shard, _server._config.bounce_request_smp_service_group,
+            [this, is, cs = cs.move_to_other_shard(), stream, permit = std::move(permit), process_fn,
              gt = tracing::global_trace_state_ptr(std::move(trace_state)),
-             cached_vals = std::move(bounce_msg->take_cached_pk_function_calls()), dialect] (cql_server& server) {
+             cached_vals, dialect] (cql_server& server) mutable {
```

Comment:
You dropped wrong `move`. The lambda is created on the src cpu, but the `move` below in `do_with` will be executed on a dst cpu and will move pointers from src to dst cpu. You need to drop the `move` below. Also does the lambda have to be mutable?

Reply:
> Hmm. does not invoke_on run the lambda on the remote cpu? if so, the lambda captures should had already been constructed at the moment of invocation?

`invoke_on` runs lambda on the remote shard, but it creates it on a local shard. So yes lambda captures are already constructed when the lambda is invoked, but they constructed on another shard.

Reply:
> in my understanding the lambda capture `cached_vals` is not a ref, but a copy, created on the src cpu, from which we are moving on the dst cpu.

This is correct. But this copy contains pointers on src cpu. When we move it (as opposite to copy) on a dst cpu we copy the pointer and then we do cross shard free of the pointer. What we should do it to move into the lamda on src and copy from the lambda on dst.

Reply:
I agree with @gleb-cloudius. If it was big and we wanted to avoid a copy we could use foreign unique ptr, so it would get freed back on the originating shard, but they way it is right now, it must not be moved on a different shard.  Previously without the `mutable` specifier, std::move didn't actually move the object content because it was const in practice, so the bug was latent.  Making the lambda mutable exposes this bug.

Reply:
> @gleb-cloudius thanks for the explanation, now I see. we still can have an optimization by using a reference in [this](https://github.com/scylladb/scylladb/pull/20493/commits/f4e4d6e09364b1c1988ecceb34af519dc572001e) commit since it introduced co_await.

Doe you mean we can capture by reference? Right.

Reply:
Yes, I meant 'capture by reference'.

Reply:
Yes. We can avoid the copy.

Reply:
Fixed in the new version.